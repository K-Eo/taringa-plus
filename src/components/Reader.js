import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Loader, Message } from 'semantic-ui-react'

import './Reader.css'

import * as actions from '../actions/reader'
import { defaultSelector, controlSelector } from '../selectors/reader'

import { toHTML } from '../BBCodeParser'

const createMarkup = (content) => ({ __html: toHTML(content) })

export class Reader extends Component {
  componentDidMount () {
    const { match } = this.props

    if (match && match.params && match.params.id) {
      this.props.fetch(match.params.id)
    }

    window.scrollTo(0, 0)
  }

  render () {
    const { control, post } = this.props

    if (!control) {
      return null
    }

    if (control.status === 'fetching') {
      return (
        <Loader active inline='centered'/>
      )
    }

    if (control.status === 'failure') {
      return (
        <Message
          icon='frown'
          header='Houston, tenemos un problema!'
          content='Ha ocurrido un error. Intenta recargar la pÃ¡gina.'
          negative
        />
      )
    }

    return (
      <div className='ui grid'>
        <div
          className='eleven wide column'
          dangerouslySetInnerHTML={createMarkup(post.body)}
        />
      </div>
    )
  }
}

Reader.defaultProps = {
  fetch: () => {}
}

const mapStateToProps = (state, props) => ({
  post: defaultSelector(state, props),
  control: controlSelector(state, props)
})

export default connect(mapStateToProps, actions)(Reader)
