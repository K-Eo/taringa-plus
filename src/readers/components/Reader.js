import React, { Component } from 'react'
import TimeAgo from 'react-timeago'
import { connect } from 'react-redux'
import {
  Comment,
  Header,
  Loader,
  Message
} from 'semantic-ui-react'

import './Reader.css'
import { esFormatter } from '../../Utils'

import Recommends from '../../recommends/components/Recommends'

import { fetch as fetchRecommends } from '../../recommends/actions'
import { fetch as fetchComments } from '../../comments/actions'
import * as actions from '../actions'

import { defaultSelector, controlSelector } from '../selectors'
import { recommendsSelector } from '../../recommends/selectors'
import { commentsSelector } from '../../comments/selectors'

import { toHTML } from '../../BBCodeParser'

const createMarkup = (content) => ({ __html: toHTML(content) })

export class Reader extends Component {
  componentDidMount () {
    const { match } = this.props

    if (match && match.params && match.params.id) {
      const id = match.params.id

      this.props.fetchPost(id)
      this.props.fetchRecommends(id)
      this.props.fetchComments(id)
    }

    window.scrollTo(0, 0)
  }

  componentWillReceiveProps (nextProps) {
    const { match } = this.props

    if (match.params.id !== nextProps.match.params.id) {
      const id = nextProps.match.params.id

      this.props.fetchPost(id)
      this.props.fetchRecommends(id)
      this.props.fetchComments(id)

      window.scrollTo(0, 0)
    }
  }

  render () {
    const {
      comments,
      control,
      dispatch,
      post,
      recommends
    } = this.props

    if (!control) {
      return null
    }

    if (control.status === 'failure') {
      return (
        <Message
          icon='frown'
          header='Houston, tenemos un problema!'
          content='Ha ocurrido un error. Intenta recargar la pÃ¡gina.'
          negative
        />
      )
    }

    return (
      <div className='ui grid'>
        <div className='eleven wide column'>
          {
            control.status === 'fetching' && <Loader active inline='centered'/>
          }
          {
            (control.status === 'success' && post) && (
              <div dangerouslySetInnerHTML={createMarkup(post.body)}/>
            )
          }
          {
            control.status === 'success' && (
              <Comment.Group>
                <Header dividing>Comentarios</Header>
                {
                  comments.map(comment => (
                    <Comment key={comment.id}>
                      <Comment.Avatar src={comment.owner.avatar.big}/>
                      <Comment.Content>
                        <Comment.Author as='a'>
                          {comment.owner.nick}
                        </Comment.Author>
                        <Comment.Metadata>
                          <TimeAgo
                            date={comment.created}
                            formatter={esFormatter}
                          />
                        </Comment.Metadata>
                        <Comment.Text>
                          {comment.body}
                        </Comment.Text>
                        <Comment.Actions>
                          <Comment.Action>Responder</Comment.Action>
                        </Comment.Actions>
                      </Comment.Content>
                    </Comment>
                  ))
                }
              </Comment.Group>
            )
          }
        </div>
        <div
          className='five wide column'
        >
          <Recommends
            dispatch={dispatch}
            {...recommends}
          />
        </div>
      </div>
    )
  }
}

Reader.defaultProps = {
  fetchPost: () => {},
  fetchRecommends: () => {},
  fetchComments: () => {},
  recommends: {
    error: '',
    posts: [],
    status: 'success'
  },
  comments: []
}

const mapStateToProps = (state, props) => ({
  comments: commentsSelector(state),
  control: controlSelector(state, props),
  post: defaultSelector(state, props),
  recommends: recommendsSelector(state)
})

const mapDispatchToProps = {
  fetchComments: fetchComments,
  fetchPost: actions.fetch,
  fetchRecommends: fetchRecommends
}

export default connect(mapStateToProps, mapDispatchToProps)(Reader)
