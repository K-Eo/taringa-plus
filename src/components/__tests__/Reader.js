import React from 'react'
import renderer from 'react-test-renderer'

import { Reader } from '../Reader'

describe('Reader', () => {
  it('renders', () => {
    const component = renderer.create(
      <Reader
        control={{ status: 'success' }}
        post={{ body: '[b]foobar[/b]' }}
      />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders nothing if no control', () => {
    const component = renderer.create(
      <Reader control={undefined}/>
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders loader when fetching', () => {
    const component = renderer.create(
      <Reader control={{ status: 'fetching' }}/>
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders message error', () => {
    const component = renderer.create(
      <Reader control={{ status: 'failure' }}/>
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('fetchs post on componentDidMount when id is given', () => {
    const match = {
      params: {
        id: 1
      }
    }

    const mock = jest.fn()

    renderer.create(
      <Reader match={match} fetchPost={mock}/>
    )

    expect(mock).toBeCalledWith(1)
  })

  it('does not fetch post on componentDidMount when id is missing', () => {
    const match = {
      params: { }
    }

    const mock = jest.fn()

    renderer.create(
      <Reader match={match} fetchPost={mock}/>
    )

    expect(mock).not.toBeCalled()
  })

  it('fetchs recommends on componentDidMount when id is given', () => {
    const match = {
      params: {
        id: 1
      }
    }

    const mock = jest.fn()

    renderer.create(
      <Reader match={match} fetchRecommends={mock}/>
    )

    expect(mock).toBeCalledWith(1)
  })

  it('does not fetch post on componentDidMount when id is missing', () => {
    const match = {
      params: { }
    }

    const mock = jest.fn()

    renderer.create(
      <Reader match={match} fetchRecommends={mock}/>
    )

    expect(mock).not.toBeCalled()
  })

  it('scrolls to top when componentDidMount', () => {
    const mock = jest.fn()

    window.scrollTo = mock

    renderer.create(
      <Reader/>
    )

    expect(mock).toBeCalledWith(0, 0)
  })
})
