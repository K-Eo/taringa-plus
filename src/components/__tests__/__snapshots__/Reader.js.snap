// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader renders 1`] = `
<div
  className="ui grid"
>
  <div
    className="eleven wide column"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<strong>foobar</strong>",
      }
    }
  />
</div>
`;

exports[`Reader renders loader when fetching 1`] = `
<div
  className="ui active centered inline loader"
/>
`;

exports[`Reader renders message error 1`] = `
<div
  className="ui icon negative message"
>
  <i
    aria-hidden="true"
    className="frown icon"
  />
  <div
    className="content"
  >
    <div
      className="header"
    >
      Houston, tenemos un problema!
    </div>
    <p>
      Ha ocurrido un error. Intenta recargar la p√°gina.
    </p>
  </div>
</div>
`;

exports[`Reader renders nothing if no control 1`] = `null`;
